### PR TITLE
GPU support for Myelin-based parser

### PR DESCRIPTION
**Pull request by ringgaard**
The Myelin-based SLING parser can now run on GPUs using the CUDA API. All the kernels are running on the GPU so very little data need to be transferred between the host and the device. The channels for the recurrent connections are also stored in GPU memory. I have added support for "fast fallback" which allows the argmax computation to run on the GPU. Then only the final prediction needs to be read from the GPU instead having to read all the 6000+ logits.

Support for CUDA in Myelin:
  * CUDA API with dynamic loading
  * PTX assembler for generating code for the CUDA kernels
  * Myelin runtime for CUDA
  * CUDA kernel for launching kernels on GPUs
  * Support for Nvidia intrincics in Express
  * CUDA kernels for expressions, matmul, etc.
  * ArgMax kernel for CPUs with AVX2
  * Channel allocation on GPU device
  * Overhead profiling
  * Tensor transfer consolidation
  * A number of additional bug fixes for Myelin

### CODE REVIEW

File: myelin/kernel/cuda-arithmetic.cc [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+    // Compute element offset.
+    ptx_decl(b64, offset);
+    comp.offset = offset;
+    ptx_emit(mul.wide.u32, offset, idx, PTXImm(comp.size));
+    ptx_decl(b64, addr);
+    comp.addr = addr;
+
+    // Allocate registers.
+    comp.reg.resize(regs);
+    std::vector<PTXReg> reg(regs);
```
Comment (rahul1980) [inline myelin/kernel/cuda-arithmetic.cc]: This vector doesn't seem to be used.

File: myelin/cuda/cuda-runtime.cc [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+CUDARuntime::~CUDARuntime() {
+  Disconnect();
+}
+
+void CUDARuntime::Connect(int device_number) {
+  // Check if device is already connected to another device.
+  if (device_ != nullptr) {
+    if (device_number != -1 && device_->number() != device_number) {
+      LOG(FATAL) << "CUDA runtime already connect to device";
+    }
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/cuda/cuda-runtime.cc]: If the runtime has already been connected to a device (i.e. device_ != nullptr) and you specify that you are willing to connect to any device (i.e. device_number == -1), then we will just let the runtime be connected to the current device.

File: myelin/cuda/cuda-runtime.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  ~CUDARuntime();
+
+  // Connect runtime to CUDA devices. If the device number is -1 the runtime
+  // tries to selected the best GPU device for computations.
+  void Connect(int device_number = -1);
+
+  // Disconnect runtime from device.
+  void Disconnect();
+
+  // Return description of the CPU device.
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/cuda/cuda-runtime.h]: Fixed.

File: myelin/kernel/cuda-arithmetic.cc (hunk 1 above)
Comment (ringgaard) [reply myelin/kernel/cuda-arithmetic.cc]: You're right. The regs are now in the comp block. Removed.

File: myelin/kernel/cuda-arithmetic.cc [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+ private:
+  const string name_;       // kernel name
+  const string operation_;  // kernel operation
+  int arity_;               // number of inputs
+};
+
+// CUDA-based argmax using reduction.
+class CUDAArgMax : public CUDAKernel {
+ public:
+  static const int MAX_BLOCK_SIZE = 512;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/kernel/cuda-arithmetic.cc]: Yes. My gtx 1080ti card seems to have 1024 as the max block size. I have changed this to read the max block size and the warp size from the device.

File: myelin/kernel/cuda-array.cc [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+// limitations under the License.
+
+#include "myelin/kernel/cuda.h"
+
+#include "myelin/cuda/cuda-kernel.h"
+
+namespace sling {
+namespace myelin {
+
+// Concatenation of input tensors along first dimension using CUDA
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/kernel/cuda-array.cc]: Fixed.

File: myelin/kernel/cuda-array.cc [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    // Get feature id.
+    ptx_decl(b64, fptr);
+    ptx->LoadTensorAddress(fptr, f);
+    ptx_decl(s32, fval);
+    ptx_emit(mad.wide.u32, fptr, fidx, PTXImm(f->stride(1)), fptr);
+    ptx_emit(ld.global.s32, fval, PTXAddr(fptr));
+
+    // Output zero if feature value is -1.
+    ptx_decl(f32, value);
+    ptx_emit(mov.f32, value, PTXFloat(0));
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/kernel/cuda-array.cc]: While that is technically true, it will still execute the same number of instructions, so it will not be faster. The instructions between ptx_if and ptx_endif will be conditionally executed based on the predicate value, but I think they will still take up an instruction cycle even when the predicate is false.

File: myelin/kernel/cuda-array.cc [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+    ptx_endif();
+
+    // Set OOV indicator to 1.0 if feature is -1.
+    ptx_decl(f32, oovval);
+    ptx_if(oov);
+    ptx_emit(mov.f32, oovval, PTXFloat(1.0));
+    ptx_else();
+    ptx_emit(mov.f32, oovval, PTXFloat(0));
+    ptx_endif();
+    ptx_emit(st.global.f32, PTXAddr(rrow, dims * sizeof(float)), oovval);
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply myelin/kernel/cuda-array.cc]: Changed sizeof(float) to R->stride(1).
